File: src/offer18-2.DeleteDuplicatedNode/delete_duplicated_node.cpp
/*
struct ListNode {
    int val;
    struct ListNode *next;
    ListNode(int x) :
        val(x), next(NULL) {
    }
};
*/
class Solution {
public:
    ListNode* deleteDuplication(ListNode* pHead)
    {
        if (pHead == nullptr || pHead->next == nullptr) return pHead;
        ListNode* newHead = new ListNode(-1);
        newHead->next = pHead;
        ListNode* pre = newHead;
        ListNode* cur = pHead;
        ListNode* next = nullptr;
        while (cur != nullptr && cur->next != nullptr){
            next = cur->next;
            if (cur->val == next->val){
                while (next != nullptr && next->val == cur->val){
                    delete cur;
                    cur = next;
                    next = next->next;
                }
                delete cur;
                cur = next;
                pre->next = cur;
            }
            else{
                pre = cur;
                cur = cur->next;
            }
        }
        return newHead->next;
    }
};

File: src/offer18-2.DeleteDuplicatedNode/delete_duplicated_node.java
/*
 public class ListNode {
    int val;
    ListNode next = null;

    ListNode(int val) {
        this.val = val;
    }
}
*/
public class Solution {
    public ListNode deleteDuplication(ListNode pHead)
    {
        ListNode pCur = pHead;
        ListNode pPre = null;
        while(pCur != null){
            ListNode pNxt = pCur.next;
            if(pNxt != null && pCur.val == pNxt.val){
                ListNode pTmp = pNxt;
                int val = pCur.val;
                while(pTmp != null && pTmp.val == val){
                    pTmp = pTmp.next;
                }
                if(pPre != null){
                    pPre.next = pTmp;
                }else{
                    pHead = pTmp;
                }
                pCur = pTmp;
            }else{
                pPre = pCur;
                pCur = pCur.next;
            }
        }
        return pHead;
    }
}